File: hola-cdn-sdk/hola_log.m
//
//  HolaCDNLog.m
//  hola-cdn-sdk
//
//  Created by alexeym on 27/07/16.
//  Copyright © 2017 hola. All rights reserved.
//

#import "hola_log.h"

@implementation HolaCDNLog

static HolaCDNLogLevel verboseLevel = HolaCDNLogLevelError;
static NSArray* verboseModules;

static dispatch_once_t onceToken;
static NSString* const defaultModules[] = {};
static int const defaultModulesCount = 0;

- (instancetype)init:(NSString*)module
{
    self = [super init];
    if (self) {
        _module = module == nil ? @"cdn" : module;

        NSNumber* instances = [HolaCDNLog instances][_module];
        if (instances == nil) {
            [[HolaCDNLog instances] setObject:[NSNumber numberWithInt:1] forKey:_module];
            _instance = 1;
        } else {
            _instance = [instances intValue]+1;
            [[HolaCDNLog instances] setObject:[NSNumber numberWithInt:_instance] forKey:_module];
        }

        dispatch_once(&onceToken, ^{
            if (verboseModules == nil) {
                [HolaCDNLog setVerboseModules:nil];
            }
        });
    }
    return self;
}

+(NSMutableDictionary*)instances {
    static NSMutableDictionary* instances = nil;
    if (instances == nil) {
        instances = [NSMutableDictionary new];
    }

    return instances;
}

+(instancetype)logWithModule:(NSString*)module {
    return [[HolaCDNLog alloc] init:module];
}

+(void) setVerboseLevel:(HolaCDNLogLevel) level {
    verboseLevel = level;
}

+(void) setVerboseModules:(NSArray*) modules {
    if (modules == nil) {
        NSArray* dm = [NSArray arrayWithObjects:defaultModules count:defaultModulesCount];
        verboseModules = dm;
        return;
    }

    verboseModules = modules;
}

-(NSString*) prefixForLevel:(HolaCDNLogLevel)level {
    NSString* levelString;

    switch (level) {
      case HolaCDNLogLevelNone:
        levelString = @"";
        break;
      case HolaCDNLogLevelDebug:
        levelString = @"DEBUG";
        break;

      case HolaCDNLogLevelInfo:
        levelString = @"INFO";
        break;

      case HolaCDNLogLevelWarning:
        levelString = @"WARN";
        break;

      case HolaCDNLogLevelError:
        levelString = @"ERROR";
        break;

      case HolaCDNLogLevelCritical:
        levelString = @"CRIT";
        break;
    }

    NSString* instanceID = _instance != 0 ? [NSString stringWithFormat:@":%d", _instance] : @"";
    return [NSString stringWithFormat:@"[%@/%@%@]", levelString, _module, instanceID];
}

-(void) rawLogWithLevel:(HolaCDNLogLevel)level andMessage:(NSString *)msg {
    if (level < verboseLevel) {
        return;
    }

    if (verboseModules == nil || [verboseModules count] == 0 || [verboseModules containsObject:_module]) {
        NSLog(@"%@ %@", [self prefixForLevel:level], msg);
    }
}

-(void) debug:(NSString*) msg {
    [self rawLogWithLevel:HolaCDNLogLevelDebug andMessage:msg];
}

-(void) info:(NSString*) msg {
    [self rawLogWithLevel:HolaCDNLogLevelInfo andMessage:msg];
}

-(void) warn:(NSString*) msg {
    [self rawLogWithLevel:HolaCDNLogLevelWarning andMessage:msg];
}

-(void) err:(NSString*) msg {
    [self rawLogWithLevel:HolaCDNLogLevelError andMessage:msg];
}

-(void) crit:(NSString*) msg {
    [self rawLogWithLevel:HolaCDNLogLevelCritical andMessage:msg];
}

@end

File: hola_cdn/hola_log.swift
//
//  hola_log.swift
//  hola_cdn
//
//  Created by norlin on 04/07/16.
//  Copyright © 2016 hola. All rights reserved.
//

import Foundation

public enum HolaCDNLogLevel: Int {
    case None = 0
    case Debug = 1
    case Info = 2
    case Warning = 3
    case Error = 4
    case Critical = 5
}

class HolaCDNLog: NSObject {
    static var level: HolaCDNLogLevel = .Error
    static func setLevel(level: HolaCDNLogLevel) {
        HolaCDNLog.level = level
    }
    
    let module: String
    
    init(module: String) {
        self.module = module
    }
    
    private func prefix(level: HolaCDNLogLevel) -> String {
        let levelString: String
        
        switch level {
        case .Debug:
            levelString = "DEBUG"
            break
        case .Info:
            levelString = "INFO "
            break
        case .Warning:
            levelString = "WARN "
            break
        case .Error:
            levelString = "ERR  "
            break
        case .Critical:
            levelString = "CRIT "
            break
        case .None:
            return ""
        }
        
        return "[\(levelString)/\(module)] "
    }
    
    private func rawLog(level: HolaCDNLogLevel, msg: String) {
        if (level.rawValue < HolaCDNLog.level.rawValue) {
            return
        }
        
        let prefix = self.prefix(level)
        
        NSLog("\(prefix)\(msg)")
    }
    
    func debug(msg: String) {
        rawLog(.Debug, msg: msg)
    }
    
    func info(msg: String) {
        rawLog(.Info, msg: msg)
    }
    
    func warn(msg: String) {
        rawLog(.Warning, msg: msg)
    }
    
    func err(msg: String) {
        rawLog(.Error, msg: msg)
    }
    
    func crit(msg: String) {
        rawLog(.Critical, msg: msg)
    }
}
